crate/memories/src/batching_layer/memory.rs

```
@@ -0,0 +1,217 @@
+                (
+                    Operation::Read((addresses, sender)),
+                    OperationResultReceiver::<M>::Read(receiver),
+                )
+            }
+            MemoryInput::Write((guard, bindings)) => {
+                let (sender, receiver) = oneshot::channel();
+                (
+                    Operation::Write(((guard, bindings), sender)),
+                    OperationResultReceiver::<M>::Write(receiver),
+                )
+            }
+        };
+
+        if let Some(ops) = self.buffer.push(operation)? {
+            self.manage(ops).await?;
+        }
+
+        Ok(match receiver {
+            OperationResultReceiver::Read(receiver) => {
+                let result = receiver.await?.map_err(MemoryBatcherError::Memory)?;
+                MemoryOutput::Read(result)
+            }
+            OperationResultReceiver::Write(receiver) => {
+                let result = receiver.await?.map_err(MemoryBatcherError::Memory)?;
+                MemoryOutput::Write(result)
+            }
+        })
+    }
+
+    async fn manage(&self, ops: PendingOperations<M>) -> Result<(), MemoryBatcherError<M>> {
+        // Assumes the vector is homogeneous, i.e. all operations are of the same type.
+        // This should be guaranteed by the buffer.
+        match ops[0] {
+            Operation::Read(_) => {
+                // Build combined address list while tracking which addresses belong to which
+                // batch.
+                let all_addresses: Vec<_> = ops
+                    .iter()
+                    .flat_map(|op| match op {
+                        Operation::Read((addresses, _)) => addresses.clone(),
+                        _ => unreachable!(
+                            "Expected all operations to be reads, reaching this statement means \
+                             the buffer has implementation flaws at the push level."
+                        ),
```

Maybe return an error instead. If you do so, you should remove the homogeneity check in `push`.

---

The new code looks more complex but it's necessary, the `unreachable!` seemed more clear and since it was (really) unreachable we could avoid the double collect (that's basically like a double iteration, right ?). Putting an error looks like it "can" happen

Here is the new committed code, please tell me what to keep
```rust
let all_addresses: Vec<_> = ops
                    .iter()
                    .map(|op| match op {
                        Operation::Read((addresses, _)) => Ok(addresses.clone()),
                        _ => Err(MemoryBatcherError::Buffering(
                            crate::batching_layer::buffer::BufferError::TypeMismatch,
                        )),
                    })
                    .collect::<Result<Vec<_>, _>>()? // This will short-circuit on first error
                    .into_iter()
                    .flatten()
                    .collect();
```

The previous code : 

```rust
let all_addresses: Vec<_> = ops
                    .iter()
                    .flat_map(|op| match op {
                        Operation::Read((addresses, _)) => addresses.clone(),
                        _ => unreachable!(
                            "Expected all operations to be reads, reaching this statement means \
                             the buffer has implementation flaws at the push level."
                        ),
                    })
                    .collect();
```

---

IMHO I vouch for the previous method, costing an O(1) and garanteeing no weird behaviors vs. the current one that's O(N) (collect) and permits a scenario with heterogenous vector, but I will let you chose

---

to avoid comments spam, the two messages above also apply for this too : 
```rust
 // Distribute results to each batch's sender.
                for (input_addresses, sender) in ops
                    .into_iter()
                    .map(|op| match op {
                        Operation::Read((addresses, sender)) => (addresses, sender),
                        _ => unreachable!(
                            "Expected all operations to be reads, reaching this statement means \
                             the buffer has implementation flaws at the push level."
                        ),
                    })
                    .rev()
```

and also this one : 

```rust
let (bindings, senders): (Vec<_>, Vec<_>) = ops
                    .into_iter()
                    .map(|op| match op {
                        Operation::Write((bindings, sender)) => (bindings, sender),
                        _ => unreachable!(),
                    })
                    .unzip();
```

same approach and same considerations